lib/kafka/consumer.rb

```diff
@@ -474,5 +501,34 @@ def fetch_batches(min_bytes:, max_bytes:, max_wait_time:, automatically_mark_as_
 
       raise FetchError, e
     end
+
+    RETRYABLE_POLL_EXCEPTIONS = [
+      HeartbeatError,
+      OffsetCommitError,
+      FetchError,
+      NotLeaderForPartition,
+      UnknownTopicOrPartition,
+      LeaderNotAvailable,
+    ]
+
+    def retry_or_raise(retries:)
+      try = 0
+      begin
+        @logger.debug "Attempt ##{try} of #{retries}"
+        yield
+      rescue *RETRYABLE_POLL_EXCEPTIONS => ex
+        if [HeartbeatError, OffsetCommitError].include?(ex.class)
+          join_group
+        elsif [FetchError, NotLeaderForPartition, UnknownTopicOrPartition].include?(ex.class)
+          @cluster.mark_as_stale!
+        elsif LeaderNotAvailable == ex.class
+          @logger.error "Leader not available; waiting 1s before retrying"
+          @cluster.mark_as_stale!
+          sleep 1
+        end
```

Comment:
This could be done more elegantly as a `case` expression:

```ruby
case ex
when HeartbeatError, OffsetCommitError
  join_group
when FetchError, NotLeaderForPartition, UnknownTopicOrPartition
  @cluster.mark_as_stale!
when LeaderNotAvailable
  # ...
end
```